### ISSUE TITLE
kdbus breaks libvirtd 1.2.17

### ISSUE THREAD

**Issue by heftig** [systemd/systemd#550]
When starting, libvirtd complains that `operation failed: network 'default' already exists with uuid 1a41bcc0-41a1-414e-b6fe-c26997680eb0`.

When starting virt-manager, libvirtd outputs a _lot_ of error messages in the form `operation failed: filter 'XXX' already exists with uuid XXX`. This also happens when stopping libvirtd.

 virt-manager shows that no virtual networks exist. Attempting to start a VM fails because the `default` network is missing.

**Comment by zonque** [comment]
Could you provide any debugging hints from `busctl monitor` or `strace`?

**Comment by heftig** [comment]
While `busctl monitor` is running, the bus seems unusable. `systemctl start` hangs and `d-feet` gets a timeout on all its attempts at introspection.

**Comment by heftig** [comment]
From `strace`, this might be interesting:

```
[pid 21601] sendmsg(13, {msg_name(0)=NULL, msg_iov(2)=[{"l\1\1\1q\0\0\0\2\0\0\0\177\0\0\0\1\1o\0\25\0\0\0/org/freedesktop/DBus\0\0\0\6\1s\0\24\0\0\0org.freedesktop.DBus\0\0\0\0\2\1s\0\24\0\0\0org.freedesktop.DBus\0\0\0\0\3\1s\0\10\0\0\0AddMatch\0\0\0\0\0\0\0\0\10\1g\0\1s\0\0", 144}, {"l\0\0\0type='signal',interface='org.freedesktop.DBus',member='NameOwnerChanged',arg0='org.fedoraproject.FirewallD1'\0", 113}], msg_controllen=0, msg_flags=0}, MSG_NOSIGNAL <unfinished ...>
[pid 21601] <... sendmsg resumed> )     = 257
[pid 21601] sendmsg(13, {msg_name(0)=NULL, msg_iov(2)=[{"l\1\1\1M\0\0\0\3\0\0\0\177\0\0\0\1\1o\0\25\0\0\0/org/freedesktop/DBus\0\0\0\6\1s\0\24\0\0\0org.freedesktop.DBus\0\0\0\0\2\1s\0\24\0\0\0org.freedesktop.DBus\0\0\0\0\3\1s\0\10\0\0\0AddMatch\0\0\0\0\0\0\0\0\10\1g\0\1s\0\0", 144}, {"H\0\0\0type='signal',interface='org.fedoraproject.FirewallD1',member='Reloaded'\0", 77}], msg_controllen=0, msg_flags=0}, MSG_NOSIGNAL <unfinished ...>
[pid 21601] <... sendmsg resumed> )     = 221
[pid 21585] recvmsg(13,  <unfinished ...>
[pid 21585] <... recvmsg resumed> {msg_name(0)=NULL, msg_iov(1)=[{"l\3\1\1/\0\0\0\377\377\377\377\177\0\0\0\5\1u\0\2\0\0\0\4\1s\0)\0\0\0org.freedesktop.DBus.Error.ServiceUnknown\0\0\0\0\0\0\0\6\1s\0\10\0\0\0:1.89159\0\0\0\0\0\0\0\0\7\1s\0\24\0\0\0org.freedesktop.DBus\0\0\0\0\10\1g\0\1s\0\0*\0\0\0Destination org.freedesktop.DBus not known\0l\3\1\1/\0\0\0\377\377\377\377\177\0\0\0\5\1u\0\3\0\0\0\4\1s\0)\0\0\0org.freedesktop.DBus.Error.ServiceUnknown\0\0\0\0\0\0\0\6\1s\0\10\0\0\0:1.89159\0\0\0\0\0\0\0\0\7\1s\0\24\0\0\0org.freedesktop.DBus\0\0\0\0\10\1g\0\1s\0\0*\0\0\0Destination org.freedesktop.DBus not known\0", 2048}], msg_controllen=0, msg_flags=MSG_CMSG_CLOEXEC}, MSG_CMSG_CLOEXEC) = 382

```

**Comment by heftig** [comment]
Fiddling with /run/libvirt and the network state got me around the network problem, but libvirt remains broken. Starting a VM causes libvirt to launch qemu and register the VM with machined, but it never advances the VM out of "shutoff" state.

**Comment by dvdhrm** [comment]
I can reproduce those `already exists with uuid` errors. The issue here is simple: libvirt installs dbus matches to get notified when firewalld reloads configuration. Unfortunately, libvirt prints a huge set of errors if you reload configuration. I don't know why, and I don't know whether that code of libvirt actually works, but that's not up to me to fix.

Instead, I looked at why libvirt reloads the configuration even though firewalld isn't even installed on my machine. As it turns out, libvirt just matches for "NameOwnerChanged" signals and handles all of them equally. This is fine on dbus1 as it has reliable bus-filters. However, kdbus does not have this. Instead, you might get signals you didn't subscribe to (still subject to policy, though). You're responsible to handle such false positives locally and discard them.

As it turns out, some applications cannot handle those. So attached is a patch that forces bus-proxyd to discard un-matched broadcasts. This silences the libvirt errors I saw (and the one you described).

As I have no clue how libvirt works, I didn't actually test whether the other issues you described still work. So would you mind testing this? (Note that the log_error() introduced here is triggered by a lot of other stuff; I will turn it into log_debug() before pushing it upstream).

```
diff --git a/src/bus-proxyd/driver.c b/src/bus-proxyd/driver.c
index 4ac955d..1cb5ea5 100644
--- a/src/bus-proxyd/driver.c
+++ b/src/bus-proxyd/driver.c
@@ -33,6 +33,7 @@
 #include "strv.h"
 #include "set.h"
 #include "driver.h"
+#include "proxy.h"
 #include "synthesize.h"

 static int get_creds_by_name(sd_bus *bus, const char *name, uint64_t mask, sd_bus_creds **_creds, sd_bus_error *error) {
@@ -70,7 +71,7 @@ static int get_creds_by_message(sd_bus *bus, sd_bus_message *m, uint64_t mask, s
         return get_creds_by_name(bus, name, mask, _creds, error);
 }

-int bus_proxy_process_driver(sd_bus *a, sd_bus *b, sd_bus_message *m, SharedPolicy *sp, const struct ucred *ucred, Set *owned_names) {
+int bus_proxy_process_driver(Proxy *p, sd_bus *a, sd_bus *b, sd_bus_message *m, SharedPolicy *sp, const struct ucred *ucred, Set *owned_names) {
         int r;

         assert(a);
@@ -189,7 +190,7 @@ int bus_proxy_process_driver(sd_bus *a, sd_bus *b, sd_bus_message *m, SharedPoli
                 if (r < 0)
                         return synthetic_reply_method_errno(m, r, NULL);

-                r = sd_bus_add_match(a, NULL, match, NULL, NULL);
+                r = sd_bus_add_match(a, NULL, match, proxy_match, p);
                 if (r < 0)
                         return synthetic_reply_method_errno(m, r, NULL);

diff --git a/src/bus-proxyd/driver.h b/src/bus-proxyd/driver.h
index b8cedf5..da3834f 100644
--- a/src/bus-proxyd/driver.h
+++ b/src/bus-proxyd/driver.h
@@ -23,5 +23,6 @@

 #include "sd-bus.h"
 #include "bus-xml-policy.h"
+#include "proxy.h"

-int bus_proxy_process_driver(sd_bus *a, sd_bus *b, sd_bus_message *m, SharedPolicy *sp, const struct ucred *ucred, Set *owned_names);
+int bus_proxy_process_driver(Proxy *p, sd_bus *a, sd_bus *b, sd_bus_message *m, SharedPolicy *sp, const struct ucred *ucred, Set *owned_names);
diff --git a/src/bus-proxyd/proxy.c b/src/bus-proxyd/proxy.c
index 189ee96..8145891 100644
--- a/src/bus-proxyd/proxy.c
+++ b/src/bus-proxyd/proxy.c
@@ -144,9 +144,17 @@ static int proxy_create_local(Proxy *p, int in_fd, int out_fd, bool negotiate_fd
         return 0;
 }

+static int proxy_match_synthetic(sd_bus_message *m, void *userdata, sd_bus_error *error) {
+        Proxy *p = userdata;
+
+        p->synthetic_matched = true;
+        return 0; /* make sure to continue processing it in further handlers */
+}
+
 /*
- * dbus-1 clients receive NameOwnerChanged and directed signals without
- * subscribing to them; install the matches to receive them on kdbus.
+ * We always need NameOwnerChanged so we can synthesize NameLost and
+ * NameAcquired. Furthermore, dbus-1 always passes unicast-signals through, so
+ * subscribe unconditionally.
  */
 static int proxy_prepare_matches(Proxy *p) {
         _cleanup_free_ char *match = NULL;
@@ -172,7 +180,7 @@ static int proxy_prepare_matches(Proxy *p) {
         if (!match)
                 return log_oom();

-        r = sd_bus_add_match(p->destination_bus, NULL, match, NULL, NULL);
+        r = sd_bus_add_match(p->destination_bus, NULL, match, proxy_match_synthetic, p);
         if (r < 0)
                 return log_error_errno(r, "Failed to add match for NameLost: %m");

@@ -189,7 +197,7 @@ static int proxy_prepare_matches(Proxy *p) {
         if (!match)
                 return log_oom();

-        r = sd_bus_add_match(p->destination_bus, NULL, match, NULL, NULL);
+        r = sd_bus_add_match(p->destination_bus, NULL, match, proxy_match_synthetic, p);
         if (r < 0)
                 return log_error_errno(r, "Failed to add match for NameAcquired: %m");

@@ -202,7 +210,7 @@ static int proxy_prepare_matches(Proxy *p) {
         if (!match)
                 return log_oom();

-        r = sd_bus_add_match(p->destination_bus, NULL, match, NULL, NULL);
+        r = sd_bus_add_match(p->destination_bus, NULL, match, proxy_match_synthetic, p);
         if (r < 0)
                 log_error_errno(r, "Failed to add match for directed signals: %m");
                 /* FIXME: temporarily ignore error to support older kdbus versions */
@@ -679,6 +687,7 @@ static int patch_sender(sd_bus *a, sd_bus_message *m) {

 static int proxy_process_destination_to_local(Proxy *p) {
         _cleanup_bus_message_unref_ sd_bus_message *m = NULL;
+        bool matched = false, matched_synthetic = false;
         int r;

         assert(p);
@@ -692,6 +701,21 @@ static int proxy_process_destination_to_local(Proxy *p) {
         }
         if (r == 0)
                 return 0;
+
+        /*
+         * Usually, we would just take any message that the bus passes to us
+         * and forward it to the local connection. However, there are actually
+         * applications that fail if they receive broadcasts that they didn't
+         * subscribe to. Therefore, we actually emulate a real broadcast
+         * matching here, and discard any broadcasts that weren't matched. Our
+         * match-handlers remembers whether a message was matched by any rule,
+         * by marking it in @p->message_matched.
+         */
+        matched = p->message_matched;
+        matched_synthetic = p->synthetic_matched;
+        p->message_matched = false;
+        p->synthetic_matched = false;
+
         if (!m)
                 return 1;

@@ -705,6 +729,15 @@ static int proxy_process_destination_to_local(Proxy *p) {
         if (r < 0)
                 return log_error_errno(r, "Failed to synthesize message: %m");

+        /* discard broadcasts that were not matched by any MATCH rule */
+        if (!matched && !sd_bus_message_get_destination(m)) {
+                if (matched_synthetic)
+                        log_error("Dropped unmatched broadcast: uid=" UID_FMT " gid=" GID_FMT" message=%s path=%s interface=%s member=%s",
+                                  p->local_creds.uid, p->local_creds.gid, bus_message_type_to_string(m->header->type),
+                                  strna(m->path), strna(m->interface), strna(m->member));
+                return 1;
+        }
+
         patch_sender(p->destination_bus, m);

         if (p->policy) {
@@ -788,7 +821,7 @@ static int proxy_process_local_to_destination(Proxy *p) {
         if (r > 0)
                 return 1;

-        r = bus_proxy_process_driver(p->destination_bus, p->local_bus, m, p->policy, &p->local_creds, p->owned_names);
+        r = bus_proxy_process_driver(p, p->destination_bus, p->local_bus, m, p->policy, &p->local_creds, p->owned_names);
         if (r == -ECONNRESET || r == -ENOTCONN)
                 return r;
         if (r < 0)
@@ -834,6 +867,13 @@ static int proxy_process_local_to_destination(Proxy *p) {
         return 1;
 }

+int proxy_match(sd_bus_message *m, void *userdata, sd_bus_error *error) {
+        Proxy *p = userdata;
+
+        p->message_matched = true;
+        return 0; /* make sure to continue processing it in further handlers */
+}
+
 int proxy_run(Proxy *p) {
         int r;

diff --git a/src/bus-proxyd/proxy.h b/src/bus-proxyd/proxy.h
index ff278a2..ccb951c 100644
--- a/src/bus-proxyd/proxy.h
+++ b/src/bus-proxyd/proxy.h
@@ -39,6 +39,8 @@ struct Proxy {

         bool got_hello : 1;
         bool queue_overflow : 1;
+        bool message_matched : 1;
+        bool synthetic_matched : 1;
 };

 int proxy_new(Proxy **out, int in_fd, int out_fd, const char *dest);
@@ -46,6 +48,7 @@ Proxy *proxy_free(Proxy *p);

 int proxy_set_policy(Proxy *p, SharedPolicy *policy, char **configuration);
 int proxy_hello_policy(Proxy *p, uid_t original_uid);
+int proxy_match(sd_bus_message *m, void *userdata, sd_bus_error *error);
 int proxy_run(Proxy *p);

 DEFINE_TRIVIAL_CLEANUP_FUNC(Proxy*, proxy_free);
```

**Comment by dvdhrm** [comment]
Ok, so I digged into `virt-manager` and I can create and start VMs with this patch applied just fine.

**Comment by heftig** [comment]
Seems to work well, thanks.

**Comment by dvdhrm** [comment]
This is fixed by #603. It's not merged, yet, but I'll close it here as the issue is now known and debugged. I'll take care it gets merged in some way.

Thanks a lot for the report!
